llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4812,15 +4812,20 @@ Instruction *InstCombinerImpl::visitFreeze(FreezeInst &I) {
   //
   // TODO: This could use getBinopAbsorber() / getBinopIdentity() to avoid
   //       duplicating logic for binops at least.
-  auto getUndefReplacement = [&I](Type *Ty) {
-    Constant *BestValue = nullptr;
-    Constant *NullValue = Constant::getNullValue(Ty);
+  auto getUndefReplacement = [&I, &AC = this->AC, &DT = this->DT](Type *Ty) {
```

Comment:
I don't really like to put the logic inside this lambda function. IIRC `getUndefReplacement` returns a **scalar** constant to fill in "holes" in a constant vector.

Reply:
Not really. The *primary* purpose is to replace a freeze(poison) fully, see the `if (match(Op0, m_Undef()))` branch below. However, it is *also* used to handle the case where we have something like `freeze <i8 0, i8 poison>`, where we're only replacing elements.

TBH I find it somewhat doubtful to use the same helper for both. I think it would make more sense to separate out the vector handling and make that one just fill the gaps with the first non-undef element. I think creating a vector splat is going to be more useful than creating a a mixed vector constant where some elements have beneficial values for some operation.

Also, I think that code is currently incorrect because it fails to handle potentially poisonous constant expressions.

Maybe we should clean up that code first.

Reply:
From a high level should my implementation plan be?

1. Simplify getUndefReplacement helper to only handle select scalars
2. Implement poisonous constant expressions
3. Implement vector splat generation